File: backend/joyagent-core/src/main/java/com/jd/genie/controller/ChatSessionController.java
package com.jd.genie.controller;

import com.jd.genie.agent.agent.AgentContext;
import com.jd.genie.model.multi.EventMessage;
import com.jd.genie.model.req.AgentRequest;
import com.jd.genie.model.response.AgentResponse;
import com.jd.genie.service.ChatSessionService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.*;

/**
 * 聊天会话控制器
 * 处理会话相关的API请求
 */
@RestController
@RequestMapping("/api/chat")
public class ChatSessionController {

    private static final Logger logger = LoggerFactory.getLogger(ChatSessionController.class);
    private final ChatSessionService chatSessionService;

    @Autowired
    public ChatSessionController(ChatSessionService chatSessionService) {
        this.chatSessionService = chatSessionService;
    }

    /**
     * 获取所有会话
     */
    @GetMapping("/sessions")
    public ResponseEntity<List<Map<String, Object>>> getAllSessions() {
        try {
            List<Map<String, Object>> sessions = chatSessionService.getAllSessions();
            return ResponseEntity.ok(sessions);
        } catch (Exception e) {
            logger.error("获取所有会话失败", e);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    /**
     * 获取指定会话
     */
    @GetMapping("/sessions/{sessionId}")
    public ResponseEntity<Map<String, Object>> getSession(@PathVariable String sessionId) {
        try {
            Optional<Map<String, Object>> session = chatSessionService.getSession(sessionId);
            return session.map(ResponseEntity::ok)
                    .orElseGet(() -> ResponseEntity.notFound().build());
        } catch (Exception e) {
            logger.error("获取会话失败: {}", sessionId, e);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    /**
     * 创建新会话
     */
    @PostMapping("/sessions")
    public ResponseEntity<Map<String, Object>> createSession(@RequestBody Map<String, Object> metadata) {
        try {
            Map<String, Object> session = chatSessionService.createSession(metadata);
            return ResponseEntity.status(HttpStatus.CREATED).body(session);
        } catch (Exception e) {
            logger.error("创建会话失败", e);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    /**
     * 删除会话
     */
    @DeleteMapping("/sessions/{sessionId}")
    public ResponseEntity<Void> deleteSession(@PathVariable String sessionId) {
        try {
            boolean deleted = chatSessionService.deleteSession(sessionId);
            return deleted ? ResponseEntity.ok().build() : ResponseEntity.notFound().build();
        } catch (Exception e) {
            logger.error("删除会话失败: {}", sessionId, e);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    /**
     * 清空所有会话
     */
    @DeleteMapping("/sessions")
    public ResponseEntity<Void> clearAllSessions() {
        try {
            chatSessionService.clearAllSessions();
            return ResponseEntity.ok().build();
        } catch (Exception e) {
            logger.error("清空所有会话失败", e);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    /**
     * 获取会话消息
     */
    @GetMapping("/sessions/{sessionId}/messages")
    public ResponseEntity<Map<String, Object>> getSessionMessages(@PathVariable String sessionId) {
        try {
            Optional<List<Map<String, Object>>> messages = chatSessionService.getSessionMessages(sessionId);
            if (messages.isPresent()) {
                Map<String, Object> response = new HashMap<>();
                response.put("code", 0);
                response.put("data", messages.get());
                response.put("message", "success");
                response.put("requestId", sessionId);
                return ResponseEntity.ok(response);
            } else {
                return ResponseEntity.notFound().build();
            }
        } catch (Exception e) {
            logger.error("获取会话消息失败: {}", sessionId, e);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    /**
     * 保存会话消息
     */
    @PostMapping("/sessions/{sessionId}/messages")
    public ResponseEntity<Void> saveSessionMessages(
            @PathVariable String sessionId,
            @RequestBody Map<String, Object> payload) {
        
        try {
            @SuppressWarnings("unchecked")
            List<Map<String, Object>> messages = (List<Map<String, Object>>) payload.get("messages");
            Map<String, Object> metadata = payload.containsKey("metadata") 
                    ? (Map<String, Object>) payload.get("metadata") 
                    : Map.of();
            
            boolean saved = chatSessionService.saveSessionMessages(sessionId, messages, metadata);
            return saved ? ResponseEntity.ok().build() : ResponseEntity.badRequest().build();
        } catch (Exception e) {
            logger.error("保存会话消息失败: {}", sessionId, e);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    /**
     * 同步本地会话到后端
     */
    @PostMapping("/sessions/{sessionId}/sync")
    public ResponseEntity<Void> syncSession(
            @PathVariable String sessionId,
            @RequestBody Map<String, Object> session) {
        
        try {
            boolean synced = chatSessionService.syncSession(sessionId, session);
            return synced ? ResponseEntity.ok().build() : ResponseEntity.badRequest().build();
        } catch (Exception e) {
            logger.error("同步会话失败: {}", sessionId, e);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }
} 

File: backend/joyagent-core/src/main/python/com/jd/genie/controller/ChatSessionController.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
聊天会话控制器
提供会话相关的API接口
"""

import json
from typing import Dict, Any, List, Optional
from fastapi import APIRouter, Request, HTTPException, Path, Body
from pydantic import BaseModel
from loguru import logger

from com.jd.genie.service.ChatSessionService import chat_session_service

router = APIRouter()


class SessionResponse(BaseModel):
    """会话响应模型"""
    code: int = 0
    data: Any = None
    message: str = "success"
    requestId: str = ""


@router.get("/sessions")
async def get_all_sessions():
    """获取所有会话"""
    try:
        sessions = chat_session_service.get_all_sessions()
        return SessionResponse(data=sessions)
    except Exception as e:
        logger.error(f"获取所有会话失败: {e}")
        return SessionResponse(
            code=500,
            message="获取会话失败",
            data=[]
        )


@router.get("/sessions/{session_id}")
async def get_session(session_id: str = Path(...)):
    """获取指定会话"""
    try:
        session = chat_session_service.get_session(session_id)
        if session:
            return SessionResponse(data=session)
        else:
            return SessionResponse(
                code=404,
                message="会话不存在",
                data=None
            )
    except Exception as e:
        logger.error(f"获取会话失败: {session_id}, {e}")
        return SessionResponse(
            code=500,
            message=f"获取会话失败: {str(e)}",
            data=None
        )


@router.post("/sessions")
async def create_session(metadata: Dict[str, Any] = Body(...)):
    """创建新会话"""
    try:
        session = chat_session_service.create_session(metadata)
        return SessionResponse(
            data=session,
            message="会话创建成功"
        )
    except Exception as e:
        logger.error(f"创建会话失败: {e}")
        return SessionResponse(
            code=500,
            message=f"创建会话失败: {str(e)}",
            data=None
        )


@router.delete("/sessions/{session_id}")
async def delete_session(session_id: str = Path(...)):
    """删除会话"""
    try:
        deleted = chat_session_service.delete_session(session_id)
        if deleted:
            return SessionResponse(message="会话删除成功")
        else:
            return SessionResponse(
                code=404,
                message="会话不存在",
                data=None
            )
    except Exception as e:
        logger.error(f"删除会话失败: {session_id}, {e}")
        return SessionResponse(
            code=500,
            message=f"删除会话失败: {str(e)}",
            data=None
        )


@router.delete("/sessions")
async def clear_all_sessions():
    """清空所有会话"""
    try:
        chat_session_service.clear_all_sessions()
        return SessionResponse(message="所有会话已清空")
    except Exception as e:
        logger.error(f"清空所有会话失败: {e}")
        return SessionResponse(
            code=500,
            message=f"清空所有会话失败: {str(e)}",
            data=None
        )


@router.get("/sessions/{session_id}/messages")
async def get_session_messages(session_id: str = Path(...)):
    """获取会话消息"""
    try:
        messages = chat_session_service.get_session_messages(session_id)
        if messages is not None:
            return SessionResponse(
                data=messages,
                requestId=session_id
            )
        else:
            return SessionResponse(
                code=404,
                message="会话不存在或没有消息",
                data=[],
                requestId=session_id
            )
    except Exception as e:
        logger.error(f"获取会话消息失败: {session_id}, {e}")
        return SessionResponse(
            code=500,
            message=f"获取会话消息失败: {str(e)}",
            data=[],
            requestId=session_id
        )


class MessagePayload(BaseModel):
    """消息请求负载"""
    messages: List[Dict[str, Any]]
    metadata: Optional[Dict[str, Any]] = None


@router.post("/sessions/{session_id}/messages")
async def save_session_messages(
    session_id: str = Path(...),
    payload: MessagePayload = Body(...)
):
    """保存会话消息"""
    try:
        saved = chat_session_service.save_session_messages(
            session_id,
            payload.messages,
            payload.metadata or {}
        )
        if saved:
            return SessionResponse(message="消息保存成功")
        else:
            return SessionResponse(
                code=400,
                message="消息保存失败",
                data=None
            )
    except Exception as e:
        logger.error(f"保存会话消息失败: {session_id}, {e}")
        return SessionResponse(
            code=500,
            message=f"保存会话消息失败: {str(e)}",
            data=None
        )


@router.post("/sessions/{session_id}/sync")
async def sync_session(
    session_id: str = Path(...),
    session: Dict[str, Any] = Body(...)
):
    """同步会话"""
    try:
        if session_id != session.get("id"):
            return SessionResponse(
                code=400,
                message="会话ID不匹配",
                data=None
            )
            
        synced = chat_session_service.sync_session(session_id, session)
        if synced:
            return SessionResponse(message="会话同步成功")
        else:
            return SessionResponse(
                code=400,
                message="会话同步失败",
                data=None
            )
    except Exception as e:
        logger.error(f"同步会话失败: {session_id}, {e}")
        return SessionResponse(
            code=500,
            message=f"同步会话失败: {str(e)}",
            data=None
        ) 